api/v1beta1/cluster_types.go

```diff
@@ -451,6 +469,77 @@ type ClusterStatus struct {
 	// ObservedGeneration is the latest generation observed by the controller.
 	// +optional
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// v1beta2 groups all the fields that will be added or modified in Cluster's status with the V1Beta2 version.
+	// +optional
+	V1Beta2 *ClusterV1Beta2Status `json:"v1beta2,omitempty"`
+}
+
+// ClusterV1Beta2Status groups all the fields that will be added or modified in Cluster with the V1Beta2 version.
+// See https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20240916-improve-status-in-CAPI-resources.md#machineset-newconditions for more context.
+type ClusterV1Beta2Status struct {
+	// conditions represents the observations of a Cluster's current state.
+	// Known condition types are Available, InfrastructureReady, ControlPlaneInitialized, ControlPlaneAvailable, WorkersAvailable, MachinesReady
+	// MachinesUpToDate, RemoteConnectionProbe, ScalingUp, ScalingDown, Remediating, Deleting, Paused.
+	// Additionally, a TopologyReconciled will be added in case the Cluster is referencing a ClusterClass / defining a managed Topology.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// controlPlane groups all the observations about Cluster's ControlPlane current state.
+	// +optional
+	ControlPlane *ClusterControlPlaneStatus `json:"controlPlane,omitempty"`
+
+	// workers groups all the observations about Cluster's Workers current state.
+	// +optional
+	Workers *WorkersStatus `json:"workers,omitempty"`
+}
+
+// ClusterControlPlaneStatus groups all the observations about control plane current state.
+type ClusterControlPlaneStatus struct {
+	// desiredReplicas is the total number of desired control plane machines in this cluster.
+	// +optional
+	DesiredReplicas int32 `json:"desiredReplicas"`
+
+	// replicas is the total number of control plane machines in this cluster.
+	// +optional
+	Replicas int32 `json:"replicas"`
+
+	// upToDateReplicas is the number of up-to-date control plane machines in this cluster. A machine is considered up-to-date when Machine's UpToDate condition is true.
```

Comment:
When would an UpToDate condition be true, or not? How would a user know what it means for an UpToDate condition to be true or not?

Reply:
We can think about swagger or about adding something to our book, but out of date comments in our API (or out of date documentation in general) is an issue we are struggling with since a long time, and I think we should really stop to do things that could make this problem worst.

Also, this is not really a problem specific of this PR only and it should be solved in a more structural way.

Reply:
Will have a think about a good way to solve this in general for the project then, otherwise PR LGTM